### DISCUSSION TITLE
Any way to get more information out of SetMatches?

### DISCUSSION THREAD

**Discussion by simonwuelker** [rust-lang/regex#805] [Q&A] [answered]
When doing something like
```rs
let re = Regex::new("[0-9]{3}-[0-9]{3}-[0-9]{4}").unwrap();
let mat = re.find("phone: 111-222-3333").unwrap();
```
`mat` contains information like the start/end index of the match.
Is there any way to get the same information when using a `RegexSet`?
As far as i can tell, when using `RegexSet` you can only tell **if** a expression matched the input, not **where**.
Why is that?

**Comment by BurntSushi** [comment, marked as answer]
> Is there any way to get the same information when using a `RegexSet`?

Have you [checked the docs](https://docs.rs/regex/0.2.5/regex/struct.RegexSet.html#limitations)?

> As far as i can tell, when using RegexSet you can only tell if a expression matched the input, not where.

Correct.

> Why is that?

See:

* https://github.com/rust-lang/regex/issues/156
* https://github.com/rust-lang/regex/issues/352
* https://github.com/rust-lang/regex/issues/259
* https://github.com/rust-lang/regex/issues/696

With all that said, I am anticipating expanding the API of `RegexSet` to permit it to report match locations. I haven't finished doing the experimentation work required to confirm it's possible, but I don't anticipate any roadblocks.

In general, reporting match locations like this requires re-arranging how the match engines work. If you don't design the match engines in the first place to support it (as is the case for this crate), then adding it after-the-fact is difficult. And moreover, it has to be added to _each_ engine.

In the course of working on #656, I have or am about to finish re-writing every internal match engine. And in so doing, I made sure to design the implementation of each to work with multiple patterns from the start.

In the end, `RegexSet` will necessarily provide overlapping matches, and this should be doable for all matching engines. However, the lower level `regex-automata` crate will provide additional multi-regex non-overlapping APIs as well.

**Reply by simonwuelker** [reply]
Okay, thanks a lot!
